### ISSUE TITLE
iPhone + Android: add immediate on-device traffic-sign recognition with national model packs

### ISSUE THREAD

**Issue by volzinnovation** [volzinnovation/youspeed.de#2]
## Problem

YouSpeed needs immediate traffic-sign recognition (TSR) from live video frames and individual captured images on both iPhone and Android. Recognition must run on-device and remain useful without a network connection.

The architecture must also support:

- national sign sets, because appearance, taxonomy, thresholds, and legal meaning differ by country;
- a posteriori, fully offline replay against the same captured inputs;
- comparison with [`Panoramax/classify_de_road_signs`](https://huggingface.co/Panoramax/classify_de_road_signs) and other open-source implementations;
- a reproducible YouSpeed-controlled training, evaluation, export, and model-release pipeline.

This issue is independent of Panoramax image capture and upload. Live TSR, model packs, replay, and training must work without a Panoramax account, upload consent, or network connection.

Related, non-blocking: Panoramax scene capture and upload is tracked in #3. Neither issue is a prerequisite for the other; each must ship and operate independently.

### 1. Recognize signs immediately on-device

Support two interchangeable recognition pipelines behind one normalized result contract:

1. **Direct multiclass detector:** one model detects and classifies national signs with minimum inference overhead.
2. **Proposal detector → classifier:** a generic or country-tuned detector locates signs, then a country-specific classifier labels stabilized crops.

```text
                                      ┌─ direct country detector ──────────┐
camera/still → preprocess + throttle ─┤                                     ├→ normalized candidate
                                      └─ sign proposal detector → crop →   │
                                                        country classifier ─┘
                                                                       ↓
                                                         temporal/spatial fusion
                                                                       ↓
                                                    live TSR UI + local observation
```

Only one pipeline runs during normal driving. Offline benchmark mode may replay the same material through several adapters.

Requirements:

- Use CameraX `ImageAnalysis` with latest-frame backpressure on Android and AVCapture video-data output on iOS.
- Keep the analysis queue at one and discard stale frames.
- Start with a measured, adaptive 2–10 FPS envelope rather than assuming a fixed rate is sufficient at all vehicle speeds.
- Analyze the useful full frame; do not permanently limit recognition to a narrow center crop.
- Track repeated detections and select the sharpest/largest crop for second-stage classification.
- Emit a prompt provisional result, then a confirmed result after the configured temporal rule.
- Require repeated evidence and appropriate heading/way/spatial stability before creating a local observation.
- Pause or degrade cleanly when backgrounded, thermally constrained, or in low-power mode.
- Do not retain continuous raw video or ordinary inference frames.
- A camera result must not silently replace map data or the current/applicable speed-limit source.

#### Still-image inference contract

- The production recognizer accepts a platform-neutral pixel-buffer/bitmap input in addition to the live-frame stream.
- The MVP user flow supports a rear-camera still captured by the app. Importing an existing image is required for diagnostic/replay tooling but does not need to be a general consumer photo-library feature.
- Normalize EXIF orientation, front/rear-camera mirroring, color space, resize/letterbox behavior, and crop coordinates before inference.
- Live frames, camera stills, and imported diagnostic fixtures use the same preprocessing version, country pack, normalized result schema, and confidence semantics.
- Still-image recognition remains local-only and creates no retained benchmark bundle unless the separate diagnostic-capture consent is enabled.

“Confirmed” means confirmed by the model’s temporal rule, not legally verified.

### 2. Support versioned national model packs

Start with Germany, but do not encode German class IDs or labels in shared app logic. A pack may contain a country-specific detector and classifier or reference a shared proposal detector.

A pack manifest should include at least:

```json
{
  "schema_version": 1,
  "pack_id": "de-speed-signs-v1",
  "countries": ["DE"],
  "pipeline": "direct_detection|proposal_classification",
  "detector": {"artifact": "...", "sha256": "..."},
  "classifier": {"artifact": "...", "sha256": "..."},
  "taxonomy_version": "tsr-semantic-v1",
  "class_mapping": "...",
  "preprocessing": "...",
  "thresholds": "...",
  "calibration": "...",
  "licenses": ["..."],
  "minimum_app_version": "..."
}
```

The classifier is optional for direct detection.

Requirements:

- CoreML and TFLite artifacts originate from the same pinned training checkpoint and manifest.
- National labels map into a stable semantic ontology, for example `maximum_speed(value, unit)`, `zone_start`, `zone_end`, `restriction_end`, `city_entry`, `temporary`, and `unknown`.
- Thresholds and score calibration are pack-specific. Do not present a raw softmax score as a probability.
- Select the active country from stable location/map context with border hysteresis and a user override. Never infer the country from the recognized sign itself.
- Permit adjacent-country packs to be cached/preloaded.
- If no valid pack exists, show TSR as unavailable rather than silently using the wrong country.
- Verify artifact hashes/signatures and compatibility before activation and preserve rollback.
- The same app build must be able to activate a new country pack without application code changes.

#### Runtime artifact contract

Each detector/classifier component must record its pinned source checkpoint and enumerate every mobile artifact explicitly. Each artifact entry includes:

- target platform and minimum OS/runtime version;
- format (`coreml`, `tflite`, or reference `onnx`), precision/quantization, input shape, and output schema;
- artifact URI/path and SHA-256;
- source-checkpoint SHA-256;
- exporter name/version and export configuration;
- calibration-dataset revision/hash when quantized;
- parity tolerance and measured reference-vs-runtime result.

The app activates only the artifact matching its platform/runtime and rejects a pack whose declared preprocessing, taxonomy, or output contract is incompatible.

### 3. Show the live classification at the top right

The top-right slot currently used for location precision becomes the TSR result on both platforms. Numeric GPS precision remains available in drive diagnostics rather than occupying the active-drive corner.

#### UI mockup

```text
Active drive — top-right slot

No recognition          Provisional              Confirmed by fusion
┌──────────────┐        ┌──────────────┐         ┌──────────────┐
│ Camera / TSR │        │ Camera · ?   │         │ Camera       │
│      —       │        │     30       │         │     30       │
│ no result    │        │ checking     │         │ detected     │
└──────────────┘        └──────────────┘         └──────────────┘

Unavailable             Unknown class
┌──────────────┐        ┌──────────────┐
│ Camera / TSR │        │ Camera       │
│      ×       │        │      ?       │
│ pack needed  │        │ unknown sign │
└──────────────┘        └──────────────┘
```

The final visual treatment should use the applicable traffic-sign pictogram/value when available, while remaining readable in the existing compact corner slot.

Requirements:

- Use “no recognition,” not “no sign”; the camera may have missed a sign.
- Provisional, confirmed, unavailable, and unknown states must differ by text/icon/shape, not color alone.
- Keep the TSR result visually distinct from the central map/current-limit display.
- Tapping the slot opens details: country/model pack and version, class/value, provisional/confirmed state, timestamp, calibrated model confidence, and “camera estimate—not legally verified”.
- If confidence is not calibrated, show a qualitative model score rather than a percentage.
- VoiceOver/TalkBack exposes the same source, class/value, state, and confidence information.
- Avoid repeated spoken announcements for changing frames; announce a transition to confirmed at most unless the user opts into more.
- Support Dynamic Type/font scaling, sufficient contrast, and the existing supported screen sizes.
- Location precision remains visible in diagnostics after it leaves the top-right slot.

### 4. Provide offline replay and comparison

Add an explicit diagnostic/benchmark capture mode, separate from ordinary TSR and Panoramax upload.

An opt-in capture bundle should contain:

- selected full frames for detector and end-to-end evaluation;
- sign crops for classifier evaluation;
- timestamps, orientation, crop coordinates, and consented location/heading/device metadata;
- live raw and normalized predictions;
- model-pack IDs and artifact hashes;
- human ground truth stored separately from predictions.

Do not retain benchmark frames during normal use. Diagnostic capture must be explicit, inspectable, deletable, and covered by a retention policy.

The replay harness must:

- run offline after model artifacts have been obtained;
- run multiple detector/classifier adapters against identical inputs;
- support reference PyTorch/ONNX plus CoreML and TFLite output comparison where applicable;
- produce a common machine-readable prediction and metrics format;
- keep classifier-on-ground-truth-crop, classifier-on-detected-crop, detector-only, and complete event-level results separate;
- pin each external model and dataset revision and checksum;
- never treat pseudo-labels as benchmark ground truth without human validation.

Initial comparison matrix:

- [Panoramax’s German road-sign classifier](https://huggingface.co/Panoramax/classify_de_road_signs) as a crop-classification baseline;
- a YouSpeed-trained YOLO nano direct detector;
- a YouSpeed-trained YOLO classifier behind a proposal detector;
- MobileNetV3 or EfficientNet-Lite as a non-YOLO classifier baseline;
- EfficientDet-Lite0 and/or YOLOX-Nano as non-Ultralytics detector baselines.

At least one additional, actual third-party open-source TSR checkpoint beyond Panoramax must be pinned, licensed, adapted, and evaluated. Merely retraining an open architecture with YouSpeed data does not satisfy the external-implementation comparison requirement. Record why any candidate could not be exported or run.

The Panoramax checkpoint is a classifier, not an end-to-end detector. Its self-reported 98.6% accuracy must not be compared directly with live full-frame TSR.

The linked [Panoramax crop dataset](https://huggingface.co/datasets/Panoramax/classified_de_road_signs) is useful bootstrap material, but its published split is not an independent acceptance set. At revision `b4856947ed7cb6312587258acc90e8cf88a4aa13`, an audit found:

- 13,287 training and 6,944 validation crops across 134 classes;
- 1,297 validation source-image IDs also present in training;
- 522 validation images byte-identical to training images.

Reproduce and document this audit against the pinned dataset revision. Rebuild leakage-safe YouSpeed evaluation splits before quoting comparison results.

Load third-party `.pt`/pickle checkpoints only inside an isolated, reproducible conversion environment. Mobile apps consume verified CoreML/TFLite artifacts, never arbitrary training checkpoints.

### 5. Build an owned training and evaluation pipeline

“Owned” means YouSpeed controls and can reproduce the taxonomy, manifests, preprocessing, split generation, configurations, training, evaluation, export, model cards, and release artifacts. Every external image/model source still requires compatible provenance and licensing.

Required stages:

1. Ingest immutable dataset revisions with source, license, and attribution manifests.
2. Normalize source labels into the versioned YouSpeed semantic taxonomy.
3. Detect exact and perceptual duplicates before splitting.
4. Group splits by physical sign, source image/sequence, route/geography, and capture period.
5. Maintain untouched country-specific field-test and out-of-distribution sets.
6. Train proposal-detector, country-classifier, and direct-detector candidates from pinned configurations.
7. Mine and review hard negatives: advertisements, windows, sign backs, adjacent roads, variable-message displays, damaged signs, and foreign signs.
8. Calibrate confidence independently for each country pack and exported runtime.
9. Export ONNX as an interchange/reference artifact and CoreML/TFLite as mobile artifacts.
10. Compare exported outputs with reference inference using shared golden fixtures and documented tolerances.
11. Generate a model card, manifest, class coverage, license/attribution record, metrics, hashes, compatibility, and rollback data.
12. Promote a country pack only after automated quality and device gates pass.

The pipeline must also version the annotation schema, label instructions, reviewer decisions, class-balance report, and label-QA results. Training runs record the code commit, container/lockfile, framework and exporter versions, random seeds, source dataset revisions, configuration, checkpoints, and experiment metrics so a released pack can be reproduced without relying on an unrecorded workstation state.

## Metrics

Report metrics by country, class, lighting/weather, sign size, and device where applicable:

- Detector: mAP50-95, small-object/proposal recall, false positives per frame.
- Classifier: macro-F1, per-class precision/recall, top-1/top-3, calibration error, unknown rejection.
- End-to-end: sign-event precision/recall, wrong-value events, misses, duplicates, false events per hour and kilometre.
- Safety-weighted errors, explicitly including severe value confusions such as `30 → 80`.
- Temporal: time to provisional/confirmed, track stability, duplicate suppression.
- Export parity: reference vs CoreML/TFLite label and score divergence.
- Device: p50/p95 latency, analyzed FPS, peak memory, battery, thermal throttling, and pack size.
- Country handling: wrong-country acceptance, missing-pack behavior, and border transitions.

## Implementation touchpoints

- Extend the existing [video TSR specification](../blob/main/docs/VIDEO_TRAFFIC_SIGN_RECOGNITION_YOLO_SPEC.md) with the recognizer contract, two-stage alternative, country packs, replay, and owned training pipeline.
- **iPhone:** AVCapture/CoreML/Vision integration; replace the top-right GPS precision badge in `MainView.swift`; extend the existing computer-vision observation modality.
- **Android:** CameraX plus LiteRT/TFLite integration; replace the top-right GPS precision badge in `ConsumerAppUi.kt`; add computer-vision observation parity.
- Keep model artifacts outside the normal source tree and distribute them through versioned, verified model packs.

## Shared camera, storage, and consent contract

When TSR and #3 coexist:

- One neutral platform camera coordinator owns session configuration and lifecycle. TSR attaches the analysis/video-data consumer; Panoramax attaches the still/photo-output consumer.
- Each feature has independent enablement, backpressure, state, and teardown. Starting or stopping one must not rebind or stop the other consumer.
- Whichever feature lands first introduces the neutral coordinator; the second extends that coordinator instead of creating a competing camera session.
- OS camera permission is not consent to retain TSR diagnostic imagery, capture Panoramax scenes, upload anything, or donate imagery for training.
- TSR benchmark bundles and Panoramax batches use separate storage roots, databases, retention controls, and consent records. Files are never aliased or reused automatically in either direction.
- Offline replay is local-only by default. Export, donation, or ingestion into training requires a separate explicit action plus a provenance/consent record.
- If Panoramax capture is active, its capture/storage indicator remains visible; TSR’s ordinary “inference retains no images” messaging must not imply that the entire app is storing nothing.

### Runtime and UI

- [ ] Both apps recognize live camera frames in airplane mode and expose the same recognizer API for an individual image.
- [ ] A rear-camera still and the same EXIF-rotated diagnostic fixture normalize into equivalent orientation/crop coordinates and the same result contract without network access.
- [ ] A provisional result reaches the top-right TSR slot within 500 ms p95 of the selected frame on agreed minimum devices under normal thermal conditions.
- [ ] A repeatedly visible sign transitions to confirmed without waiting for drive completion; the rule and measured latency are documented.
- [ ] Latest-frame backpressure prevents an inference backlog during a 30-minute drive test.
- [ ] Ordinary TSR retains neither raw video nor benchmark imagery.
- [ ] Creating, inspecting, exporting, and deleting a diagnostic bundle is covered by explicit consent/retention tests; no bundle uploads or enters training implicitly.
- [ ] No detection automatically changes map data or authoritative/current-limit state.
- [ ] The top-right slot implements no-recognition, provisional, confirmed, unavailable, and unknown behavior.
- [ ] No state relies on color alone; VoiceOver/TalkBack reads source, value/class, state, and calibrated confidence.
- [ ] The confirmed state and details identify the result as a camera estimate, not a legally verified limit.
- [ ] Numeric location precision remains available in drive diagnostics.

### Models and countries

- [ ] The same app build activates at least two fixture country packs without code changes.
- [ ] Missing, incompatible, corrupted, wrong-country, and border-transition cases fail safely.
- [ ] Direct and two-stage implementations emit the same normalized semantic result contract.
- [ ] The initial German pack covers the agreed MVP speed-sign/zone taxonomy with explicit `unknown`/`other`.
- [ ] Raw scores and calibrated confidence are not conflated.
- [ ] Every activated CoreML/TFLite artifact declares platform, precision, input/output contract, checksums, source checkpoint, exporter version, calibration revision, and passing parity result.

### Benchmark and pipeline

- [ ] One reproducible workflow builds a leakage-safe dataset release, trains at least one YouSpeed detector and classifier, evaluates them, and exports CoreML/TFLite artifacts.
- [ ] The split checker fails on source/sign-instance, sequence, exact-image, or configured perceptual duplication across splits.
- [ ] The replay harness benchmarks the pinned Panoramax classifier and other configured implementations against identical data.
- [ ] At least one additional pinned third-party open-source TSR checkpoint—not only a YouSpeed-trained open architecture—is evaluated or has a documented, reproducible incompatibility result.
- [ ] Reports keep ground-truth-crop classification, detected-crop classification, detection, and end-to-end metrics separate.
- [ ] Reference/CoreML/TFLite parity tests pass documented tolerances.
- [ ] Every released pack contains provenance, class coverage, evaluation, license/attribution, hashes, compatibility, and rollback metadata.
- [ ] Public/default enablement remains blocked until country-specific field, accuracy, false-event, latency, battery, and thermal gates are recorded and passed.

### Combined-mode compatibility

- [ ] With #3 enabled, live TSR analysis and cadence-based full-resolution still capture share one camera session for a representative 30-minute drive without session rebinding or inference backlog.
- [ ] A Panoramax still capture cannot be triggered by a TSR event, and stopping either feature does not stop the other.
- [ ] The effect of concurrent still capture on TSR p95 latency, analyzed FPS, memory, battery, and thermal behavior is measured and remains within documented release gates.
- [ ] TSR diagnostic bundles and Panoramax batches remain in separate stores and cannot be silently reused, uploaded, exported, or donated across consent boundaries.
- [ ] Combined-mode UI and accessibility clearly disclose when Panoramax is retaining full-resolution images even though ordinary TSR inference is ephemeral.

## Risks

- Two-stage detection/classification has cascading errors; always report proposal recall and compare end-to-end against direct detection.
- Small or brief signs require adaptive sampling, tracking, full-frame coverage, and dedicated small-object evaluation.
- Wrong-road association requires heading/way/spatial fusion and an advisory live display.
- Country semantics require explicit versioned mappings rather than assuming visually similar signs are equivalent.
- Border selection requires location hysteresis, manual override, adjacent-pack tests, and a safe unavailable state.
- Model confidence must be calibrated and labeled as model confidence.
- Dataset/model licenses and attribution remain binding even though YouSpeed itself is AGPL.
- Third-party checkpoints and remotely delivered artifacts require revision pinning, isolated conversion, integrity verification, and rollback.

## Out of scope

- Panoramax authentication, image capture for contribution, or image upload.
- Cloud inference or a network dependency for live TSR.
- Automatic OSM edits.
- Treating a camera result as a legally authoritative speed limit.
- Continuous raw-video persistence.
- Shipping every national sign set in the first release.
